### ISSUE TITLE
Provision pods immediately for faster node auto-provisioning

### ISSUE THREAD

**Issue by gregwebs** [pingcap/tidb-operator#355]
Currently when a new tidb cluster is created, new PD pods are created immediately. Then TiKV once PD is ready and TiDB once TiKV is ready.

Node auto provisioning looks at pending pods. Currently, the TiDB pods won't show up for a while, and their nodes are not auto-provisioned until then.

I would like to work out a design for a mode to create all the pods immediately on install (I don't care about upgrade). The pods would be initially created so that they are not schedulable (I think that means they have to request a resource that is not available?). Or they could be created so that they are scheduleable, but only start running a pause container until their dependencies are ready.

It is possible that this is generally useful for a user of tidb-operator to understand if they have capacity for all the pods that will be created.

**Comment by gregwebs** [comment]
We do already create the services ahead of time.

**Comment by gregwebs** [comment]
For over-provisioning, the cluster autoscaler [recommends using pause pods for over-provisioning](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-can-i-configure-overprovisioning-with-cluster-autoscaler).

It could be possible for us to deploy pause containers ahead of the actual containers.

**Comment by gregwebs** [comment]
There are two possibilities
1) separate pause pods
2) init containers

We could have a standard init container that blocks waiting for an http request or some other signal to stop.

**Comment by gregwebs** [comment]
We have our own custom scheduler, so it is possible for the operator to create the pods annotated as not-scheduleable (which the custom scheduler will see) and then later remove that annotation. I am not sure if this will work with the cluster autoscaler though. @aylei perhaps you can do some manual tests to find out?

**Comment by gregwebs** [comment]
But we may want to support running without our custom scheduler? Additionally we may want to get pods scheduled onto nodes immediately before there are any new nodes added to a cluster.
That would mean that having an init container is preferable.

**Comment by gregwebs** [comment]
The initcontainer approach generally inspects K8s service or pods to see if they are ready. So we can have an initContainer for TiKV that checks to see if PD is ready. And TiDB can have an init container to check that PD & TiKV are ready (or PD says TiKV is ready, etc). However, there may be a difference between our readiness probes for the container and actually checking that the cluster is properly formed and tidb-operator agrees that the service is fully ready? In that case we might want to query tidb-operator instead, but that would create an additional security burden to authorize the request in that direction (which could be across namespaces)i.

**Comment by gregwebs** [comment]
I am closing this in favor of #469 which is a proposal for launching with an init container.
